multimodal/src/autogluon/multimodal/problem_types.py

```diff
@@ -0,0 +1,221 @@
+    is_classification: bool = False  # Whether the problem is a classification problem
+    experimental: bool = False  # Indicate whether the problem is experimental
+
+    # The collection of modality types the problem supports.
+    # Multiple column types may be parsed into the same modality. For example
+    #   IMAGE, IMAGE_PATH, IMAGE_BYTEARRAY --> IMAGE
+    # It will be used to analyze the dataframe and detect the columns.
+    supported_modality_type: Set[str] = field(default_factory=set)
+
+    # The collection of label column types the problem supports.
+    supported_label_type: Optional[Set[str]] = None
+
+    # The modalities that have to appear in the table.
+    force_exist_modality: Optional[Set[str]] = None
+
+    # The fallback label type of the problem
+    _fallback_label_type: Optional[str] = None
+
+    @property
+    def fallback_label_type(self):
+        if self._fallback_label_type is None and len(self.supported_label_type) == 1:
+            return next(iter(self.supported_label_type))
+        else:
+            return self._fallback_label_type
+
+
+# Classification: Arbitrary combination of image, text, tabular data --> categorical value
+PROBLEM_TYPES_REG.register(
+    CLASSIFICATION,
+    ProblemTypeProperty(
+        name=CLASSIFICATION,
+        supported_modality_type={IMAGE, IMAGE_BYTEARRAY, TEXT, CATEGORICAL, NUMERICAL},
+        supported_label_type={CATEGORICAL},
+        is_classification=True,
+    ),
+)
+PROBLEM_TYPES_REG.register(
+    BINARY,
+    ProblemTypeProperty(
+        name=BINARY,
+        supported_modality_type={IMAGE, IMAGE_BYTEARRAY, TEXT, CATEGORICAL, NUMERICAL},
+        supported_label_type={CATEGORICAL},
+        is_classification=True,
+    ),
+)
+PROBLEM_TYPES_REG.register(
+    MULTICLASS,
+    ProblemTypeProperty(
+        name=MULTICLASS,
+        supported_modality_type={IMAGE, IMAGE_BYTEARRAY, TEXT, CATEGORICAL, NUMERICAL},
+        supported_label_type={CATEGORICAL},
+        is_classification=True,
+    ),
+)
+
+# Regression: Arbitrary combination of image, text, tabular data --> numeric value
+PROBLEM_TYPES_REG.register(
+    REGRESSION,
+    ProblemTypeProperty(
+        name=REGRESSION,
+        supported_modality_type={IMAGE, IMAGE_BYTEARRAY, TEXT, CATEGORICAL, NUMERICAL},
+        supported_label_type={NUMERICAL},
+    ),
+)
+
+# Object detection: image --> bounding boxes
+PROBLEM_TYPES_REG.register(
+    OBJECT_DETECTION,
+    ProblemTypeProperty(
+        name=OBJECT_DETECTION,
+        support_zero_shot=True,
+        supported_modality_type={IMAGE},
+        supported_label_type={ROIS},
+        force_exist_modality={IMAGE},
+    ),
+)
+
+# Matching: text <--> text, image <--> image, text <--> image
+PROBLEM_TYPES_REG.register(
+    TEXT_SIMILARITY,
+    ProblemTypeProperty(
+        name=TEXT_SIMILARITY,
+        support_zero_shot=True,
+        is_matching=True,
+        supported_modality_type={TEXT},
+        supported_label_type={CATEGORICAL, NUMERICAL},
```

Comment:
For matching problems, labels can be absent. Will it support this case?

Reply:
For those case, we can treat the label type as BINARY. But we need to submit a follow-up PR to refactor the code-base to better use these flags.

Reply:
The logic of no label is different from `binary` label. We'd better distinguish between them.

Reply:
Actually there are multiple concepts: modality type, column type and problem type. This is the first PR to split the problem type into a separate file. Need to submit follow-up PRs.

Reply:
Matching itself also needs major refactor so I haven’t touched it extensively